src/firmware/app/control/robot_controller.c

```diff
@@ -0,0 +1,73 @@
+#include "firmware/app/control/robot_controller.h"
+
+#include <assert.h>
+#include <math.h>
+#include <stdbool.h>
+
+#include "firmware/app/control/trajectory_planner.h"
+#include "firmware_new/boards/frankie_v1/io/drivetrain.h"
+
+struct RobotController
+{
+    VelocityTrajectory_t* trajectory;
+    size_t num_trajectory_elements;
+    size_t current_trajectory_index;
+    Drivetrain_t* drivetrain;
+};
+
+RobotController_t* app_robot_controller_create(Drivetrain_t* drivetrain)
+{
+    // Initialize members to zero and point to the drivetrain parameter
+    RobotController_t* controller = (RobotController_t*)malloc(sizeof(RobotController_t));
+    controller->current_trajectory_index = 0;
+    controller->num_trajectory_elements  = 0;
+    controller->trajectory               = NULL;
+    controller->drivetrain               = drivetrain;
+
+    return controller;
+}
+
+void app_robot_controller_destroy(RobotController_t* controller)
+{
+    free(controller);
+}
+
+void app_robot_controller_updateTrajectory(RobotController_t* controller,
+                                           VelocityTrajectory_t* trajectory,
+                                           size_t num_elements)
+{
+    // Update the local values to the new ones
+    controller->num_trajectory_elements = num_elements;
+    controller->trajectory              = trajectory;
+
+    // Now that there is a new trajectory, reset the current tracking index
+    controller->current_trajectory_index = 0;
+}
+
+static void io_robot_controller_sendGlobalSpeedToDrivetrain(RobotController_t* controller,
+                                                            float robot_orientation_rad,
+                                                            Matrix global_speeds)
```

Comment:
Why are we passing in `global_speeds` as a matrix here? I suppose vector's are matrices, but I really think passing in `x_velocity`, `y_velocity`, `theta_velocity` as independent args would be much more user-friendly.

Reply:
I used matrices here so I can make use of the matrix library. If we want I can break it up into 3 separate parameters

Reply:
I think from an interface perspective separate names parameters are much more clear and easy to use. But by all means leverage matrices in the implementation